python/lsst/qa/explorer/postprocess.py

```diff
@@ -0,0 +1,299 @@
+                filt: HSC-G
+                dataset: meas
+            cmodel_magDiff:
+                functor: MagDiff
+                args:
+                    - modelfit_CModel
+                    - base_PsfFlux
+                filt: HSC-G
+            gauss_magDiff:
+                functor: MagDiff
+                args:
+                    - base_GaussianFlux
+                    - base_PsfFlux
+                filt: HSC-G
+            count:
+                functor: Column
+                args:
+                    - base_InputCount_value
+                filt: HSC-G
+            deconvolved_moments:
+                functor: DeconvolvedMoments
+                filt: HSC-G
+                dataset: forced_src
+        flags:
+            - calib_psfUsed
+            - merge_measurement_i
+            - merge_measurement_r
+            - merge_measurement_z
+            - merge_measurement_y
+            - merge_measurement_g
+            - base_PixelFlags_flag_inexact_psfCenter
+            - detect_isPrimary
+
+    The names for each entry under "func" will become the names of columns in the
+    output dataset.  All the functors referenced are defined in `lsst.qa.explorer.functors`.
+    Positional arguments to be passed to each functor are in the `args` list,
+    and any additional entries for each column other than "functor" or "args" (e.g., `'filt'`,
+    `'dataset'`) are treated as keyword arguments to be passed to the functor initialization.
+
+    The "flags" entry is shortcut for a bunch of `Column` functors with the original column and
+    taken from the `'ref'` dataset.
+
+    Note, if `'filter'` is provided as part of the `dataId` when running this task (even though
+    `deepCoadd_obj` does not use `'filter'`), then this will override the `filt` kwargs
+    provided in the YAML file, and the calculations will be done in that filter.
+
+    This task uses the `lsst.qa.explorer.postprocess.PostprocessAnalysis` object
+    to organize and excecute the calculations.
+
+
+    """
+    _DefaultName = "Postprocess"
+    ConfigClass = PostprocessConfig
+
+    inputDataset = 'deepCoadd_obj'
+
+    @property
+    def outputDataset(self):
+        raise NotImplementedError('Subclass must define "outputDataset" attribute')
+
+    @classmethod
+    def _makeArgumentParser(cls):
+        """Create a suitable ArgumentParser.
+
+        """
+        parser = ArgumentParser(name=cls._DefaultName)
+        parser.add_id_argument("--id", cls.inputDataset,
+                               ContainerClass=ExistingCoaddDataIdContainer,
+                               help="data ID, e.g. --id tract=12345 patch=1,2")
+        return parser
+
+
+    def runDataRef(self, patchRef):
+        """Do calculations; write result
+        """
+        parq = patchRef.get()
+        dataId = patchRef.dataId
+        df = self.run(parq, dataId)
+        self.write(df, patchRef)
+        return df
+
+    def run(self, parq, dataId):
```

Comment:
But I require it in the run function, so how would that work?

Reply:
We're trying to get all Gen2 stuff out of `run()` (@TallJimbo knows more)  Most run methods take a dataId so that the logger can print out the dataId info in the log... helps to know which `dataId` is failing!  So our compromise was to make `dataId` an optional keyword argument. Here's a suggestion for the inserting the patch ref.  It doesn't do the `filt` though. I guess I don't understand what the dataId's filter is for. It's typicallly 'merged' for your `deepCoadd_obj` table,  for `src` it'll be 'HSC-G' or 'HSC-R'. If the filter is specified in both the `parq` AND the yaml `funcs` definitions, what does the `filt` do in PostProcessAnalysis?

```
    def runDataRef(self, patchRef):
        parq = patchRef.get()
        translationDefinition = yaml.load(open(self.config.functorFile)
        df = self.run(parq, translationDefinition, dataId=patchRef.dataId)
        self.write(df, patchRef)
        return df

    def run(self, parq, translationDef, dataId=None):
        funcs = CompositeFunctor.from_yaml(translationDef)
        df = PostprocessAnalysis(parq, funcs, filt=filt).df
        if dataId is not None:
            for key, value in dataId.items():
                df[key] = value
        return df
```